src/session.rs

```diff
@@ -0,0 +1,159 @@
+///
+/// The callback should return != 0 to indicate that a value was retrieved
+/// and written in its entirety into `buf`.
+///
+/// When `remove_after` is != 0, the returned data needs to be removed
+/// from the store.
+///
+/// NOTE: the passed in `key` and `buf` are only availabe during the
+/// callback invocation.
+/// NOTE: callbacks used in several sessions via a common config
+/// must be implemented thread-safe.
+pub type rustls_session_store_get_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_session_store_userdata,
+        key: *const rustls_slice_bytes,
+        buf: *mut u8,
+        count: size_t,
+        remove_after: c_int,
+        out_n: *mut size_t,
+    ) -> c_int,
+>;
+
+pub(crate) type SessionStoreGetCallback = unsafe extern "C" fn(
+    userdata: rustls_session_store_userdata,
+    key: *const rustls_slice_bytes,
+    buf: *mut u8,
+    count: size_t,
+    remove_after: c_int,
+    out_n: *mut size_t,
+) -> c_int;
+
+/// Prototype of a callback that can be installed by the application at the
+/// `rustls_server_config` or `rustls_client_config`. This callback will be
+/// invoked by a TLS session when a TLS session has been created and an id
+/// for later use is handed to the client/has been received from the server.
+/// `userdata` will be supplied as provided when registering the callback.
+///
+/// The callback should return != 0 to indicate that the value has been
+/// successfully persisted in its store.
+///
+/// NOTE: the passed in `key` and `val` are only availabe during the
+/// callback invocation.
+/// NOTE: callbacks used in several sessions via a common config
+/// must be implemented thread-safe.
+pub type rustls_session_store_put_callback = Option<
+    unsafe extern "C" fn(
+        userdata: rustls_session_store_userdata,
+        key: *const rustls_slice_bytes,
+        val: *const rustls_slice_bytes,
+    ) -> c_int,
+>;
+
+pub(crate) type SessionStorePutCallback = unsafe extern "C" fn(
+    userdata: rustls_session_store_userdata,
+    key: *const rustls_slice_bytes,
+    val: *const rustls_slice_bytes,
+) -> c_int;
+
+pub(crate) struct SessionStoreBroker {
+    pub userdata: rustls_session_store_userdata,
+    pub get_cb: SessionStoreGetCallback,
+    pub put_cb: SessionStorePutCallback,
+}
+
+impl SessionStoreBroker {
+    pub fn new(
+        userdata: rustls_session_store_userdata,
+        get_cb: SessionStoreGetCallback,
+        put_cb: SessionStorePutCallback,
+    ) -> Self {
+        SessionStoreBroker {
+            userdata,
+            get_cb,
+            put_cb,
+        }
+    }
+
+    fn retrieve(&self, key: &[u8], remove: bool) -> Option<Vec<u8>> {
+        let key: rustls_slice_bytes = key.into();
+        // TODO: we need a buffer where th client can store the retrieved
+        // session value. What size should it have? 10k seems excessive...
+        let mut data: Vec<u8> = vec![0 as u8; 10 * 1024];
+        let buffer = data.as_mut_slice();
```

Comment:
Vec has [`as_mut_ptr`](https://doc.rust-lang.org/std/vec/struct.Vec.html#method.as_mut_ptr), so we can skip this `let` and just do `data.as_mut_ptr()` / `data.len()` below.